tools/audio/extraction/disassemble_sequence.py

```diff
@@ -0,0 +1,1325 @@
+    return (x & (sgn - 1)) - (x & sgn)
+
+class MMLArg:
+    def __init__(self, disas):
+        self.value = self.read(disas)
+
+    def read(self, disas):
+        raise NotImplementedError()
+
+    def analyze(self, disas):
+        pass
+
+    def emit(self, disas):
+        return str(self.value)
+
+class MMLArgBits(MMLArg):
+    def read(self, disas):
+        return disas.read_bits(type(self).NBITS)
+
+class ArgU8(MMLArg):
+    def read(self, disas):
+        return disas.read_u8()
+
+class ArgU4x2(MMLArg):
+    def read(self, disas):
+        return disas.read_u8()
+
+    def emit(self, disas):
+        return f"{(self.value >> 4) & 0xF}, {self.value & 0xF}"
+
+class ArgSeqId(ArgU8):
+    def emit(self, disas):
+        return disas.all_seq_names[self.value]
+
+class ArgFontId(ArgU8): # TODO
+    def read(self, disas):
+        return disas.read_u8()
+
+class ArgPitchU8(ArgU8):
+    def emit(self, disas):
+        return f"PITCH_{pitch_names[self.value]}"
+
+class ArgS8(MMLArg):
+    def read(self, disas):
+        return sign_extend(disas.read_u8(), 8)
+
+class ArgU16(MMLArg):
+    def read(self, disas):
+        return disas.read_u16()
+
+class ArgS16(MMLArg):
+    def read(self, disas):
+        return sign_extend(disas.read_u16(), 16)
+
+class ArgHex8(ArgU8):
+    def emit(self, disas):
+        return f"0x{self.value:02X}"
+
+class ArgHex16(ArgU16):
+    def emit(self, disas):
+        return f"0x{self.value:04X}"
+
+class ArgBitField16(ArgU16):
+    def emit(self, disas):
+        return bin(self.value)
+
+class ArgInstr(ArgU8):
+    def emit(self, disas):
+        builtins = {
+            126 : "FONTANY_INSTR_SFX",
+            127 : "FONTANY_INSTR_DRUM",
+            128 : "FONTANY_INSTR_SAWTOOTH",
+            129 : "FONTANY_INSTR_TRIANGLE",
+            130 : "FONTANY_INSTR_SINE",
+            131 : "FONTANY_INSTR_SQUARE",
+            132 : "FONTANY_INSTR_NOISE",
+            133 : "FONTANY_INSTR_BELL",
+            134 : "FONTANY_INSTR_8PULSE",
+            135 : "FONTANY_INSTR_4PULSE",
+            136 : "FONTANY_INSTR_ASM_NOISE",
+        }
+        if self.value in builtins:
+            return builtins[self.value]
+
+        # Check against first font only, this is fine for 99% of cases since most sequences use just one font
+        font0 : AudiobankFile = disas.used_fonts[0]
+
+        if self.value in font0.instrument_index_map:
+            name = f"SF{font0.bank_num}_{font0.instrument_name(self.value)}"
+        else:
+            print(f"Invalid instrument sourced from {font0.name}: {self.value}")
+            name = f"{self.value} /* invalid instrument */"
+        return name
+
+class ArgVar(MMLArg):
+    def read(self, disas):
+        ret = disas.read_u8()
+        if ret & 0x80:
+            ret = ((ret << 8) & 0x7F00) | disas.read_u8()
+            if ret < 128 and disas.insn_begin not in disas.force_long:
+                print(f"Unnecessary use of long immediate encoding @ 0x{disas.insn_begin:X}: {ret}")
```

Comment:
In non-handwritten sequences this only happens in MM. In particular MM has 3 warnings in non-handwritten sequences:
```
[Sequence_90] Unnecessary use of long immediate encoding @ 0x38: 15
[Sequence_116] Invalid instrument sourced from Soundfont_36: 15
[Sequence_127] Invalid instrument sourced from Soundfont_37: 15
```
Both OoT and MM have unnecessary use of long encodings in their handwritten sequences.

Reply:
Interesting, I wonder if that indicates that there were indeed separate mnemonics for this originally, or if the encoding should be part of the mnemonic in some way (like the .s in sqrt.s I guess). Maybe it would be better to do that too instead of having the reassembler guess an encoding?

Reply:
I think it's better to do it as it is now since it's more user-friendly to let the assembler decide and just handle the special cases as-needed, but I think you're right and they probably didn't have it set up this way originally.

Reply:
Are users going to write these by hand anyway? I've been imagining that they use some kind of compiler/converter from some other format like MIDI.

Or, I suppose there could be variants like
```asm
notedv.l  ; long encoding
notedv.s  ; short encoding
notedv    ; psuedo-instruction that expands to either
```
which is not too different from how it is now, but there'd be 1 less hack and it's clearer that it's not a real instruction when the assembler decides for `notedv`.

(I'm not insisting on changing this btw, I don't think it's worth it at this point. Just spitballing)

Reply:
It's true that most won't be writing these by hand, and converters from MIDI would be used mostly. However there are cases where you do edit these by hand such as the handwritten sequences, particularly sequence 0 that implements all the sound effects. The assembler should be able to provide convenience features for these cases.

I don't mind changing the syntax to something like this, although the .s variant would never be used since the disassembler should just output the pseudo-instruction everywhere it can (so that readers of the extracted sequences don't get into bad habits)

Reply:
Hm I'm not sure it's clearly better, and I think it's a pretty big change with a fair amount of design work (which instructions need these variants? is `.s` really the right suffix or does that make it seem like a MIPS float?) which tbh I don't really want to be responsible for, given that I've been exposed to sequence assembly for about 1 hour now